packages/deployments/contracts/contracts/nomad-xapps/contracts/promise-router/PromiseRouter.sol

```diff
@@ -154,23 +211,58 @@ contract PromiseRouter is Version0, Router {
 
     bytes32 transferId = _msg.transferId();
     address callbackAddress = _msg.callbackAddress();
-    bytes memory data = _msg.returnCallData();
+    bool success = _msg.returnSuccess();
+    bytes memory data = _msg.returnData();
 
-    //TODO process callback
+    //store Promise message
+    promiseMessages[transferId] = _message;
 
     // emit Receive event
-    emit Receive(_originAndNonce(_origin, _nonce), _origin, transferId, callbackAddress, data);
+    emit Receive(_originAndNonce(_origin, _nonce), _origin, transferId, callbackAddress, success, data, _message);
   }
 
-  function process(bytes32 transactionId, bytes memory _message) internal {
-    // Should parse out the return data and callback address from message
-    // Should execute the callback() function on the provided Callback address
-    // Should enforce relayer is whitelisted by calling local connext contract
+  function process(bytes32 transferId) public payable nonReentrant {
+    // parse out the return data and callback address from message
+    bytes memory _message = promiseMessages[transferId];
+    if (_message.length == 0) revert PromiseRouter__process_InvalidMessage();
```

Comment:
Don't we need to encode and hash to check message length?

Reply:
this is more of an empty / not null check